v0/destinations/sendgrid/transform.js

```diff
@@ -0,0 +1,163 @@
+/* eslint-disable object-shorthand */
+const { EventType } = require("../../../constants");
+const {
+  CustomError,
+  getValueFromMessage,
+  getErrorRespEvents,
+  getSuccessRespEvents,
+  removeUndefinedAndNullValues,
+  defaultRequestConfig,
+  defaultPostRequestConfig,
+  isEmptyObject,
+  extractCustomFields,
+  constructPayload
+} = require("../../util");
+const {
+  ENDPOINT,
+  TRACK_EXCLUSION_FIELDS,
+  MIN_POOL_LENGTH,
+  MAX_POOL_LENGTH,
+  trackMapping
+} = require("./config");
+const {
+  payloadValidator,
+  isValidEvent,
+  createList,
+  createMailSettings,
+  createTrackSettings,
+  generatePayloadFromConfig,
+  requiredFieldValidator
+} = require("./util");
+
+const trackResponseBuilder = (message, { Config }) => {
+  let event = getValueFromMessage(message, "event");
+  if (!event) {
+    throw new CustomError("event is required for track call", 400);
+  }
+  event = event.trim().toLowerCase();
+  if (!isValidEvent(Config, event)) {
+    throw new CustomError("event not configured on dashboard", 400);
+  }
+  let payload = {};
+  payload = constructPayload(message, trackMapping);
+  if (!payload.personalizations) {
+    if (Config.mailFromTraits) {        // if enabled then we look for email in traits and if found we create personalizations object 
+      const email = getValueFromMessage(message, [
```

Comment:
should we also look for the email field in properties if not found in traits? wdyt?

Reply:
If we do that then basically it will mean that personalization field can be created from our side irrespective of whether properties field is empty or not. We can do this.